benchmarks/spiegelman_et_al_2016_gcubed/spiegelman_material.cc

```diff
@@ -0,0 +1,716 @@
+                              MaterialModel::MaterialModelOutputs<dim> &out) const;
+
+        /**
+         * Return whether the model is compressible or not.  Incompressibility
+         * does not necessarily imply that the density is constant; rather, it
+         * may still depend on temperature or pressure. In the current
+         * context, compressibility means whether we should solve the continuity
+         * equation as $\nabla \cdot (\rho \mathbf u)=0$ (compressible Stokes)
+         * or as $\nabla \cdot \mathbf{u}=0$ (incompressible Stokes).
+         *
+         * This material model is incompressible.
+         */
+        virtual bool is_compressible () const;
+
+        virtual double reference_viscosity () const;
+
+        virtual double reference_density () const;
+
+        static
+        void
+        declare_parameters (ParameterHandler &prm);
+
+        virtual
+        void
+        parse_parameters (ParameterHandler &prm);
+
+      private:
+
+        double reference_T;
+        std::vector<double> reference_T_list;
+
+
+        /**
+         * Whether or not to use the deviator of the strain-rate in the
+         * viscosity calculation.
+         */
+        bool use_deviator_of_strain_rate;
+
+        /**
+         * Defining a minimum strain rate stabilizes the viscosity calculation,
+         * which involves a division by the strain rate. Units: $1/s$.
+         */
+        std::vector<double> min_strain_rate;
+        std::vector<double> min_visc;
+        std::vector<double> max_visc;
+        std::vector<double> veff_coefficient;
+        double ref_visc;
+        double reference_compressibility;
+        std::vector<double> ref_visc_list;
+
+        std::vector<double> thermal_diffusivity;
+        std::vector<double> heat_capacity;
+
+
+        std::vector<double> densities;
+        std::vector<double> thermal_expansivities;
+
+        std::vector<double> cohesion;
+        std::vector<double> phi;
+
+        std::vector<double> constant_viscosity;
+        std::vector<double> stress_exponent;
+
+        unsigned int n_fields;
+
+        /**
+         * This parameter is used in the averaging of the  of the different compositions.
+         * It is implemented as a generalized mean, where p < -1000 is the minimum, p ==-1
+         * is the harmonic average, p == 0 is the geometric average, p == 1 is the arithmetic
+         * average and p > 1000 is the maximum.
+         */
+        double viscosity_averaging_p;
+
+        /**
+         * This value determines whether to use the analytical or the
+         * finite difference derivative for the Newton solver. Although
+         * the analytical derivative should be much faster to compute, the
+         * finite difference derivative can be useful when changing the
+         * rheology.
+         */
+        bool use_analytical_derivative;
+
+        //optimize variables
+        std::vector<double> sin_phi;
+        std::vector<double> cos_phi;
```

Comment:
Given your comment in the other place where I had a question: Can you move these two variables right below the `phi` vector above, and add a comment above `phi` of the sort "The friction angles phi for each of the compositional fields. Since we use their sines and cosines a lot, we only compute these values (for each of the phis) once and cache them."

This way, the purpose of these variables are clear.